DataDog Agent v6

rvl opened the pull request:
###### Motivation for this change

I wanted to store the `api_key` setting in a file, saw that datadog-agent v6 supports using an environment variable, then foolishly attempted packaging datadog-agent v6.

They decided to rewrite most parts of the agent in Go. The new version is here: https://github.com/DataDog/datadog-agent

The NixOS module supports both versions using the `service.dd-agent.package` option.

###### Things done

It's hard to automatically test this because it depends on a proprietary service.
I have tested a basic setup using a nixops VM and a DataDog trial account and this works.
If someone could help test more elaborate configs, that would help.

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

/cc @thoughtpolice @domenkozar @mbbx6spp @Fuuzetsu @bflyblue @coretemp

rvl committed:
datadog-agent: init at 6.1.4

DataDog have adopted a subtle naming convention:

 - dd-agent refers to the v5 Python implementation
 - datadog-agent refers to the v6 golang implementation

rvl committed:
nixos/dd-agent: Fix dodgy temp files

DataDog was dropping PID files and and python pickle files in
/tmp. Move these to private directories as a precaution.

Mic92 commented:
Will the old version be supported for a while?

rvl commented:
@Mic92 Yes I assume it will be supported by DataDog for a while because v6 still lacks features that v5 has ([changes](https://github.com/DataDog/datadog-agent/blob/master/docs/agent/changes.md), [missing_features](https://github.com/DataDog/datadog-agent/blob/master/docs/agent/missing_features.md)). Also users might need time to update any custom Python check scripts that they have.

coretemp commented on pkgs/tools/networking/dd-agent/6.nix:
```diff
@@ -0,0 +1,74 @@
+
+  # DataDog use paths relative to the agent binary, so fix these.
+  postPatch = ''
+    sed -e "s|PyChecksPath =.*|PyChecksPath = \"$bin/${python.sitePackages}\"|" \
```
I expect this to do the same and it is simpler (no escaping). Apply in multiple places. 
`sed -e 's|PyChecksPath =.*|PyChecksPath = "$bin/${python.sitePackages}"|' \`

coretemp commented:
@rvl This looks exactly like what I had in mind when I opened the issue. I hope it will be merged soon.

Mic92 commented on nixos/modules/services/monitoring/dd-agent/dd-agent.nix:
```diff
@@ -188,50 +257,71 @@ in {
+        description = "Datadog agent monitor";
+      } // (if newDog then {
+          preStart = ''
+            chown datadog /etc/dd-agent
```
This is not done recursively?

Mic92 commented on nixos/modules/services/monitoring/dd-agent/dd-agent.nix:
```diff
@@ -188,50 +257,71 @@ in {
+        serviceConfig = {
+          ExecStart = "${cfg.package}/bin/dogstatsd";
+          Type = "forking";
+          PIDFile = "/tmp/dogstatsd.pid";
```
`/tmp/` is a strange place for this. I hope statsd checks that no other user has created a malicious symlink there.

coretemp replied:
There are established locations for this stuff indeed. `/tmp` is not the right location.

For my use case this kind of security is not important, but for some people it might be.

rvl commented:
OK thanks @Mic92 I have fixed those issues. Also I noticed that the [delightful v5 dd-agent code](https://github.com/DataDog/dd-agent/blob/5.11.x/utils/pidfile.py#L17) apart from putting PID files in `/tmp`, also drops Python pickle files, which is reason enough to enable the systemd `PrivateTmp` option.

In no hurry to merge this until it gets some testing from datadog users.

Mic92 commented:
I think it is used in the core infrastructure of NixOS itself.

tazjin commented:
I haven't tested this on an actual machine yet, but from a superficial look I think this is missing the inclusion of the [core integrations](https://github.com/DataDog/integrations-core). The agent's go rewrite only covers some of the core functionality, most checks are still run via Python.

If you look at the derivation output of the V5 `dd-agent` you'll notice that the files from the integrations repository are included there. Without them the V6 agent will be able to collect basic metrics, but fail on configurations that require the integrations to be present.

Datadog's upstream packages (for Debian & co) include these in the agent package and they also note that they're included in the documentation.

I've previously read through the upstream build process that they use to create the deb & RPM packages and it's ... complicated, to say the least. I'm not really familiar with the [Python ecosystem](https://xkcd.com/1987/) but if someone wants to take a look at this together, please ping me :)

rvl committed:
pythonPackages.uuid: init at 1.30

rvl committed:
datadog-integrations-core: init at git-2018-05-27

rvl committed:
nixos/datadog-agent: add module

This is the new v6 version of datadog-agent.

The old v5 module is kept as dd-agent.

rvl commented:
Thanks @tazjin - I have added the python core integrations.

I have also decided to not make the dd-agent module work with both v5 and v6.
Instead, dd-agent will be for v5 and a new module datadog-agent will be for v6.
This allows us to make the configuration more extendable but not break the old dd-agent module API.

I have kept the v5 dd-agent temp files fix which @Mic92 pointed out.

tazjin commented:
@rvl Awesome work, I'll try it in one of our environments with a bunch of custom integration configs and report back. Probably won't get around to it before Tuesday though.

xeji commented:
The merge of #41035 created a merge conflict, please rebase.

rvl commented:
OK thanks @xeji I have rebased.

Now that there are separate NixOS modules for v5 and v6, I am more keen to get this reviewed and merged. The testing I have done is with nixops and virtualbox, with the basic host dashboard metrics, and the postgresql integration.

cransom commented:
I have a build failure when I build from the PR as well as via a `nox-review`:
```
Collecting requests (from datadog-checks-base==1.2.2)
  Could not find a version that satisfies the requirement requests (from datadog-checks-base==1.2.2) (from versions: )
No matching distribution found for requests (from datadog-checks-base==1.2.2)
```

I added `requests` to the datadog-checks-base derivation and it builds and works.

rvl commented:
Thanks @cransom - the missing requests dependency must have been caused by rebasing onto latest master. Fixed now.

I'm currently running this code on my linode with the nginx and postgresql integrations. It seems to be fine.

tazjin commented:
I've tested this on some of our test machines with simple integrations (`nginx`, `http`) enabled. Things work fine, good work! :+1: 

Planning to test it with some `process` & `postgresql` configs, too, but I'll add some comments in the diff first.

tazjin commented on nixos/modules/services/monitoring/datadog-agent.nix:
```diff
@@ -0,0 +1,205 @@
+      type = types.package;
+    };
+
+    apiKeyFile = mkOption {
```
API-wise we'd prefer if it was up to the user to determine whether they're fine with writing the key into a store path file or not. Moving config over from V5 is easily fixed with `builtins.toFile` but I thought I'd point it out anyways

tazjin commented on nixos/modules/services/monitoring/datadog-agent.nix:
```diff
@@ -0,0 +1,205 @@
+    target = "datadog-agent/" + name;
+  }) (filterAttrs (name: conf: conf != null) entries);
+
+  etcfiles = makeConfigDir
```
The module in its current form lets users specify the configuration in the structure expected by Datadog directly (i.e. there's no "fancy" translations from a setting to monitor a list of URLs to `http` checks) - that's good because it doesn't introduce another extra layer that needs to be learned.

However, I think it'd be nicer if arbitrary check configurations could be added more easily instead of having a fixed set of supported check configurations here. Currently one needs to link them via `environment.etc` in addition and then extend the restart trigger of the agent's unit.

A suggestion would be to accept an attribute set of configurations & their structure, for example like this:

```nix
services.datadog-agent.configs = {
  http_check = {
    init_config = null;
    instances = [
      # ... and so on
    ];
  };

  disk = {
    # you get the gist of it
  };
};
```

The straightforward way to implement it would be to link those to `conf.d/${key}.yaml`, but I see that some checks have their own `.d` folder - is that required? If yes, the config attribute set could probably use the whole relative path from `conf.d/` as the base as its key.

Does this make sense? :)

tazjin replied:
> is that required?

The answer is no. Datadog say that they prefer the `.d`-folder approach, but both work - so it's just a matter of choosing one.

tazjin commented on pkgs/tools/networking/dd-agent/integrations-core.nix:
```diff
@@ -0,0 +1,66 @@
+
+  });
+
+in fix' (extends overrides packages)
```
If I read this correctly the use of `fix'` is mostly to refer to `datadog_checks_base` recursively, right?

If I wanted to extend this with, say, the `ntp` integration, it isn't immediately obvious how I could reference `buildIntegration` when passing in `overrides`.

If there's a way to easily "enable" building of more integrations from the `integrations-core` repository I think it'd be great if this file had some comments explaining the reasoning and a usage example.

Maybe an alternative way is again to resort to a simple attribute set argument where `key :: IntegrationName`, where that `IntegrationName` is the `pname` of an integration in the `integrations-core` repository and `value:: [Package] -> [Package]`, which would practically be a filter for the integrations Python dependencies, for example:

```nix
pkgs.datadog-integrations-core {
  # core integrations built in addition to the defaults:
  coreIntegrations = {
    "ntp" = (pythonPkgs: []); # no extra dependencies
    "riak" = (pythonPkgs: [ pythonPkgs.httplib2 ]); # one extra dependency
  };

  # plus similar to the existing overrides arbitrary new integrations
  # (e.g. user-supplied ones):
  extraIntegrations = [ my-great-integration ];
};
```

Which could then be processed like so for the `coreIntegrations`: 

```nix
integrations = [ disk network postgres nginx mongo ] ++
  lib.mapAttrs (pname: fdeps: buildIntegration {
    inherit pname;
    propagatedBuildInputs = [ datadog_checks_base ] ++ (fdeps python.pkgs);
  }) cfg.coreIntegrations;
```

With `datadog_checks_base` declared in the `let`-binding outside of the returned attribute set (and maybe also inherited into it for convenience when building custom integrations).

tazjin commented on nixos/modules/services/monitoring/datadog-agent.nix:
```diff
@@ -0,0 +1,205 @@
+    collector_log_file  = "/var/log/datadog/collector.log";
+    forwarder_log_file  = "/var/log/datadog/forwarder.log";
+    dogstatsd_log_file  = "/var/log/datadog/dogstatsd.log";
+    pup_log_file        = "/var/log/datadog/pup.log";
```
One more thought: Most service logs on NixOS go into `journald` due to the use of systemd. Having services write logs onto disk sort of "unexpectedly" can be a bit surprising.

I don't have a strong opinion on this but it may be sensible to configure these logs to go into `journald` instead.

coretemp commented:
I would like to see this merged into master. It can always be improved later. The benefits of moving to v6 are much larger than the downside of not having complete perfection.

domenkozar commented:
+1, would love to get this in :)

rvl commented:
Yes OK let's get this in.

* I will fix the log file/journal thing that @tazjin noted.

* Making integrations more extensible should be a future PR. For this PR I just want to have the integrations which are available in the dd-agent v5 NixOS module.

* With `integrations-core.nix` I wanted to make a Python package set with DataDog integrations layered on top. Does this look right, or is there a simpler way to do it?

FRidh commented on pkgs/tools/networking/dd-agent/6.nix:
```diff
@@ -0,0 +1,76 @@
+    cp -R $src/pkg/status/dist/templates $bin/share/datadog-agent
+
+    wrapProgram "$bin/bin/agent" \
+        --prefix PYTHONPATH : "$bin/${python.sitePackages}"
```
Set is recommended over prefix.

coretemp commented:
@rvl What's the status here? I would really like to see this in 18.09.

coretemp commented:
@rvl Please also make sure that this won't suffer from https://github.com/NixOS/nixpkgs/issues/44362.

tazjin committed:
nixos/datadog-agent: Refactor to allow arbitrary check configs

Refactors the datadog-agent (i.e. V6) module to let users configure
arbitrary checks, not just a limited set, without having to resort to
linking the files manually and updating the systemd unit.

Checks are now configured via a `services.datadog-agent.checks` option
which takes an attribute set in which the keys refer directly to
Datadog check names, and the values are attribute sets representing
Datadog's configuration structure.

With this mechanism users can configure arbitrary integrations, for
example for the `ntp`-check, simply by saying:

    services.datadog-agent.checks.ntp = {
      init_config = null;
      # ... other check configuration options as per Datadog
      # documentation
    };

The previous check-specific configuration options for non-default
checks have been removed. Disk & network check configuration options
have been kept rather than making them a `default`-value of the
`checks`-option because they will be overridden by user-configurations
in that case.

Relates to NixOS/nixpkgs#40399.

tazjin commented:
@rvl @coretemp I have started fixing the things I pointed out above in my branch [feat/datadog-agent-v6](https://github.com/tazjin/nixpkgs/tree/feat/datadog-agent-v6). Ideally @rvl could cherry-pick these changes from my branch.

Changes so far (updating this continously):

* [Refactored the check configuration](https://github.com/tazjin/nixpkgs/commit/131accb37f102fa187b282b5def804d639bf8adf) let users provide configure arbitrary checks

tazjin committed:
dd-agent: Simplify inclusion of additional Datadog core integrations

Refactors the process used to build the Datadog core integrations to
be more easily extensible with integrations other than the ones built
and installed by default.

Documentation has been added in relevant parts of the module to
describe how the process works.

As a high-level overview:

The `datadog-integrations-core` attribute in the top-level package set
now accepts an extra parameter.

This parameter is an attribute set where each key is the name of a
Datadog integration as it appears in Datadog's integrations-core
repository[1], and the value is a function that receives the Python
package set and returns the required dependencies of this integration.

For example:

    datadog-integrations-core {
      ntp = (ps: [ ps.ntplib ]);
    };

This would build the default integrations and, additionally, the `ntp`
integration.

To support passing the modified Python environment to the
datadog-agent itself, the `python` key has been moved inside of the
derivation which means that it will be made overridable.

This relates to NixOS/nixpkgs#40399.

[1]: https://github.com/DataDog/integrations-core

tazjin committed:
nixos/datadog-agent: Add option to configure datadog integrations

Introduces an option `services.datadog-agent.extraIntegrations` that
can be set to include additional Datadog agent integrations from the
integrations-core repository.

Documentation and an example is provided with the change.

Relates to NixOS/nixpkgs#40399

tazjin commented:
The changes I made in my branch now cover most of the functionality we're interested in. As for the logging changes and what @FRidh pointed out, those should probably be amended into the `init` commits instead of added on at the end.

@rvl: Do you have some time for the last remaining bits on this, or would you prefer if I opened a separate PR from my branch (of course keeping all your original commits)?

@coretemp @domenkozar Is one of you available for a quick review of my changes?

rvl commented:
Hello @tazjin - thanks for helping. I have made the fixes for logging, merge conflict, and wrapProgram locally, but wasn't able to get to testing it. I can push those tomorrow when I get a little time.

I have given you write access to my nixpkgs fork so that you can push to the branch. However it might be easier if we merge this basic version without all the bells and whistles, then add the fully extensible checks/configuration in a new PR.

tazjin commented:
@rvl Thank you! :tada: If you don't mind, I can make those same changes and force-push a rebased version onto your branch.

I actually think the "bells and whistles" are slightly more straightforward than a version with specific handling for some things. Do you see a disadvantage in including the changes? That way we don't break the module's interface in `master` so much (it *is* unstable, but still :wink:)

coretemp commented on nixos/modules/services/monitoring/datadog-agent.nix:
```diff
@@ -0,0 +1,205 @@
+          rm -f /etc/datadog-agent/auth_token
+        '';
+        script = ''
+          export DD_API_KEY=$(head -n1 ${cfg.apiKeyFile})
```
`head -n 1` is POSIX correct. `head -n1` is not (it just happens to work).

tazjin commented:
I am force-pushing over the current tip (`e9b06d0efcffff97a8427137451d5b814f05f183`) on `rvl/datadog-agent` with a rebase, logging fixes, the change @coretemp pointed out, the `makePrefix` thing and my changes.

The configuration is tested on a few machines with both custom check config and extra integrations. From my side this should be done, but I haven't explicitly tested whether this is affected by #44362 - maybe @coretemp can do that, or maybe we can look at it later.

coretemp commented:
@tazjin Looking at it later is the option that I prefer. It's just that I raised it, because the costs of fixing something earlier is generally lower. However, I consider getting it into 18.09 even with this minor bug to be even more important. I don't plan to fix it in the next 4 months for example.

rvl commented:
@tazjin Sorry for the delay. I have been busy moving house.

I have just tested your fixes and refactors with nginx and postgres checks.

It looks good and works well.

Thanks.

tazjin commented:
@rvl No worries, hope the move went well! :) And thanks!

Alright, @Mic92 - is there anything missing to get this in?

disassembler commented:
LGTM! Thanks for the contribution @rvl @tazjin!

tazjin commented:
Thanks @disassembler! :1st_place_medal:

tazjin commented:
@coretemp Yes, you can do something roughly like this:

```nix
{ config, pkgs, ... }:

let

# Download nixpkgs from master (you may want to pin a commit, or maybe
# the unstable channel, instead):
unstableSrc = fetchTarball "https://github.com/NixOS/nixpkgs/archive/master.zip";

# The above gives is the raw (unimported) Nix source, which we'll need
# in a second, but we also need to import the actual package set:
unstable = import unstableSrc {};

in {
  # Import the module itself from unstable:
  imports = [
    "${unstableSrc}/nixos/modules/services/monitoring/datadog-agent.nix"
  ];

  config = {
    # Pick the relevant packages into your local overrides:
    nixpkgs.config.packageOverrides = base: base // {
      inherit (unstable) datadog-agent datadog-integrations-core;
    };

    # And use the module:
    services.datadog-agent = {
      enable     = true;
      apiKeyFile = whatever;
      tags       = [ "my-env" ];
    };
  };
}
```

I wrote this from memory so it may be slightly off, but that's roughly how we do it atm.

coretemp commented:
Thanks @tazjin. 

I am using the following (I don't like a moving target):

```
  # repository containing datadog package version 6
  src_datadog_6 = bootstrap.fetchFromGitHub {
    owner = "NixOS";
    repo  = "nixpkgs";
    rev = "b7e50d52f018d0670614a58851b4fc2c0f93818b";
    sha256 = "0gwgflq4v9zyflpfrqdnzqwk9sfvn7aqy39m0vwbrhvlz3h8dwb2";
  };
```

coretemp commented:
@tazjin What do you do about?

```
 (datadog_agent.go:129 in LogMessage) | (disk.py:105) | Unable to get disk metrics for /sys/kernel/debug/tracing: [Errno 13] Permission denied: '/sys/kernel/debug/tracing'
```

Also, I am not sure why, but despite the module running fine, the Host map in Datadog doesn't show the machine anymore after the upgrade.

coretemp commented:
@tazjin I found a more serious limitation:

I added `datadog-agent` to `environment.systemPackages`, but I am not able to execute 

`datadog-agent` per https://docs.datadoghq.com/agent/troubleshooting/#send-a-flare. (i.e. it's not on the PATH). I think the package should be added to `environment.systemPackages` in the module with some additional modifications (perhaps the "core" part).

coretemp commented:
@tazjin When I use `/nix/store/*-datadog-agent-6.1.4-bin/bin/agent flare` sending a flare also doesn't work. What is the interface you have in mind for sending such flares?

tazjin commented:
@coretemp Working on something else right now, can take a look at this tomorrow, but some small comments:

> the Host map in Datadog doesn't show the machine anymore after the upgrade

Depending on how your hosts were set up they may end up with new hostnames. This doesn't happen in my environments because we explicitly configure the hostnames used in Datadog, so I'm not sure if there is a difference in behaviour between this and the V5 module. Will check tomorrow.

> I think the package should be added to `environment.systemPackages` in the module

I'm not sure if it's common or expected for modules to install CLI tools system-wide, but I guess it makes sense here because of flares and retrieving the agent status. This should be a tiny change.

> sending a flare also doesn't work

What sort of error do you get?

coretemp commented:
@tazjin There is no error. I am certain that if you try to send a flare, that it will also not work.
